src/inmanta/agent/agent.py

```diff
@@ -398,41 +466,42 @@ def reload(
 
         **This method should only be called under critical_ratelimiter lock!**
         """
-
         # First determined if we should start and if the current run should be resumed
         if not self.finished():
             # we are still running
-            if self.is_repair:
-                # now running repair
-                if is_repair:
-                    # repair restarts repair
-                    self.logger.info("Terminating run '%s' for '%s'", self.reason, reason)
-                else:
-                    # increment interrupts repair
-                    self.logger.info("Interrupting run '%s' for '%s'", self.reason, reason)
-                    self._resume_reason = "Restarting run '%s', interrupted for '%s'" % (self.reason, reason)
+            assert self.running is not None
+            # Get correct action
+            response = deploy_response_matrix[
+                (self.running.is_full_deploy, self.running.is_periodic, new_request.is_full_deploy, new_request.is_periodic)
+            ]
+            # Execute action
+            if response == DeployRequestAction.terminate:
+                self.logger.info("Terminating run '%s' for '%s'", self.running.reason, new_request.reason)
+            elif response == DeployRequestAction.defer:
+                self.logger.info("Deferring run '%s' for '%s'", new_request.reason, self.running.reason)
+                self.deferred = new_request
```

Comment:
What if we already have a request queued? Same for the interrupt action. I think it may be safe because we only ever defer/interrupt full deploys, therefore we can just replace them (possibly excepting the message). But if this is the case, we should probably make it an invariant of the matrix and add a comment here as to why it's safe to replace `self.deferred`.

Reply:
added docs and a testcase